crates/fuzzing/src/oracles.rs

```diff
@@ -722,13 +722,135 @@ pub fn differential_wasmi_execution(wasm: &[u8], config: &crate::generators::Con
     }
 
     if &wasmi_buf[..] != &wasmtime_slice[..] {
-        show_wat();
+        show_wat(wasm);
         panic!("memory contents are not equal");
     }
 
     Some(())
 }
 
+/// Perform differential execution between Wasmtime and the official WebAssembly
+/// specification interpreter.
+///
+/// May return `None` if we early-out due to a rejected fuzz config.
+pub fn differential_spec_execution(wasm: &[u8], config: &crate::generators::Config) -> Option<()> {
+    crate::init_fuzzing();
+    debug!("WAT: {}", wasm2wat(wasm));
+    debug!("config: {:#?}", config);
+
+    let wasmtime_vals = run_in_wasmtime(wasm, config, &[]);
+    debug!("Wasmtime returned: {:?}", wasmtime_vals);
+    let spec_vals = wasm_spec_interpreter::interpret(wasm, vec![]);
+    debug!("spec interpreter returned: {:?}", &spec_vals);
+
+    // Match a spec interpreter value against a Wasmtime value. Eventually this
+    // should support references and `v128` (TODO).
+    fn matches(spec_val: &wasm_spec_interpreter::Value, wasmtime_val: &wasmtime::Val) -> bool {
+        match (spec_val, wasmtime_val) {
+            (wasm_spec_interpreter::Value::I32(a), wasmtime::Val::I32(b)) => a == b,
+            (wasm_spec_interpreter::Value::I64(a), wasmtime::Val::I64(b)) => a == b,
+            (wasm_spec_interpreter::Value::F32(a), wasmtime::Val::F32(b)) => *a as u32 == *b,
+            (wasm_spec_interpreter::Value::F64(a), wasmtime::Val::F64(b)) => *a as u64 == *b,
+            (_, _) => false,
+        }
+    }
+
+    match (&spec_vals, &wasmtime_vals) {
+        // Compare the returned values, failing if they do not match.
+        (Ok(spec_vals), Ok(wasmtime_vals)) => {
+            let all_match = spec_vals
+                .iter()
+                .zip(wasmtime_vals)
+                .all(|(s, w)| matches(s, w));
+            if !all_match {
+                show_wat(wasm);
+                panic!(
+                    "Values do not match: wasmi returned {:?}; wasmtime returned {:?}",
+                    spec_vals, wasmtime_vals
+                );
+            }
+        }
+        // If both sides fail, skip this fuzz execution.
+        (Err(spec_error), Err(wasmtime_error)) => {
+            // TODO It is unclear how this None is handled--it indicates that
+            // there were issues instantiating or running the Wasm module.
```

Comment:
Yeah, that's what I mean by "issues ... running the Wasm module." And it's actually rather unfortunate that we spend time on those cases because they aren't super useful, IMHO. I checked today and about ~50% of cases are getting to this `Err - Err` state.